charms/istio-gateway/src/charm.py

```diff
@@ -162,5 +220,68 @@ def __init__(self, msg, status_type=StatusBase):
         self.status = status_type(self.msg)
 
 
+def valid_annotations(annotations: Dict[str, str]) -> bool:
+    """Return True if the annotations pass a series of checks, False otherwise.
+
+    Validate the annotations against the following checks:
+    * Length
+    * Annotation syntax
+    * Reserved prefixes
+
+    Args:
+      annotations (Dict[str,str]): dictionary with annotations
+    """
+    for key, value in annotations.items():
+        return validate_annotation_key(key) and validate_annotation_value(value)
+    return True
+
+
+def is_qualified_name(value: str) -> bool:
+    """Check if a value is a valid Kubernetes qualified name."""
+    parts = value.split("/")
+    if len(parts) > 2:
+        return False  # Invalid if more than one '/'
+
+    if len(parts) == 2:  # If prefixed
+        prefix, name = parts
+        if not prefix or not DNS1123_SUBDOMAIN_PATTERN.match(prefix):
+            return False
+    else:
+        name = parts[0]  # No prefix
+
+    if not name or len(name) > 63 or not QUALIFIED_NAME_PATTERN.match(name):
+        return False
+
+    return True
+
+
+def validate_annotation_value(value: str) -> bool:
+    """Validate the annotation value."""
+    if not ANNOTATION_VALUE_PATTERN.match(value):
+        logger.error(
+            f"Invalid annotation value: '{value}'. Must follow Kubernetes annotation syntax."
+        )
+        return False
+
+    return True
+
+
+def validate_annotation_key(key: str) -> bool:
+    """Validate the annotation key."""
+    if len(key) > 253:
+        logger.error(f"Invalid annotation key: '{key}'. Key length exceeds 253 characters.")
+        return False
+
+    if not is_qualified_name(key.lower()):
+        logger.error(f"Invalid annotation key: '{key}'. Must follow Kubernetes annotation syntax.")
+        return False
+
+    if key.startswith(("kubernetes.io/", "k8s.io/")):
```

Comment:
suggestion: maybe we can move constants to the top of file

Reply:
Maybe tuble like ("kubernetes.io/", "k8s.io/") can be on top ... but if we dont use it e.g. in tests its fine to skip this